tests/map_test_case.h

```diff
@@ -0,0 +1,341 @@
+ *   * (optionally) apply transformations (`transpose`, `reflect_*`) can be done at any point after
+ *        `setup` and `expected_results` initialization. However, note, transformations might change
+ *        the results of getter methods.
+ *        Alternatively, use `generate_transform_combinations` inside `catch2` test to automatically
+ *        generate and test all transformations.
+ *   * use `map_tiles` variants to execute a function for each tile
+ *   * use function builders from `map_test_case_common` namespace to construct desired behavior
+ *        to use with `for_each_tile`
+ *   * use `map_test_case_common::printers` to print common debug information about the map area where
+ *        `map_test_case` is mapped to
+ */
+class map_test_case
+{
+    public:
+        // struct that is supplied as an argument to the per-tile functions (see below)
+        struct tile {
+            // current char from the `setup`
+            char sc;
+            // current char from `expected_results`
+            char ec;
+            // current point in map coords
+            tripoint p;
+            // current point in local coords [0..width) [0..height)
+            point p_local;
+        };
+
+        /**
+         *  Specifies the input 2d layout as a char matrix.
+         */
+        std::vector<std::string> setup;
+        /**
+         *  Specifies the expected results as a char matrix
+         */
+        std::vector<std::string> expected_results;
+
+        /**
+         * Char (must be uniquely present in the map_test_case::setup) that defines how the
+         * `map_test_case` coordinates are mapped onto map coordinates.
+         * When `anchor_char` is not set, top left corner (0,0) of the `setup` is used as an anchor.
+         *
+         * When using any `for_each_tile` variant, calculated (map) coordinates (`tile::p`) are such that
+         * `anchor_char` point within `setup` is aligned with the map coords stored in `anchor_map_pos`.
+         */
+        cata::optional<char> anchor_char = cata::nullopt;
+
+        /**
+         *  The 'anchor' coordinates on the map (in map coordinates).
+         *  For example, it can be the center of the map or player's coordinates.
+         *  This test will align `anchor_char` (in `setup`) with given map coordinates.
+         */
+        tripoint anchor_map_pos = tripoint_zero;
+
+        /**
+         * Invokes callback for each tile of the input map.
+         * @param callback
+         */
+        void for_each_tile( const std::function<void( tile )> &callback );
+
+        /**
+         * Invokes callback for each tile of the input map.
+         * (parameter R needed to accept functions with any return type, return value is ignored)
+         * @param callback
+         */
+        template<typename R>
+        void for_each_tile( const std::function<R( tile )> &callback );
```

Comment:
I don't think this extra overload is needed; a function returning anything should be implicitly convertible to a `std::function` returning `void`.  See e.g. https://godbolt.org/z/xfz4Eo.

Reply:
You are right, I leaved just the `void` version.